consensus/contracts_test.go

```diff
@@ -0,0 +1,331 @@
+	segment, hashes, err := hash.BuildReaderProof(bytes.NewReader(simpleFile), numSegments, segmentIndex)
+	if err != nil {
+		t.Fatal(err)
+	}
+	sp := StorageProof{
+		ContractID: txn.FileContractID(0),
+		Segment:    segment,
+		HashSet:    hashes,
+	}
+	txn = Transaction{
+		StorageProofs: []StorageProof{sp},
+	}
+
+	return
+}
+
+// testContractCreation adds a block with a file contract to the state and
+// checks that the contract is accepted.
+func testContractCreation(t *testing.T, s *State) {
+	txn := contractTxn(t, s, 2, 25*1000)
+	b, err := mineTestingBlock(s.CurrentBlock().ID(), Timestamp(time.Now().Unix()), nullMinerPayouts(s.Height()+1), []Transaction{txn}, s.CurrentTarget())
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = s.AcceptBlock(b)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Check that the contract made it into the state.
+	_, exists := s.openContracts[txn.FileContractID(0)]
+	if !exists {
+		t.Error("file contract not found found in state after being created")
+	}
+}
+
+// testMissedProof creates a contract but then doesn't submit the storage
+// proof.
+func testMissedProof(t *testing.T, s *State) {
+	// Get the transaction with the contract that will not be fulfilled.
+	txn := contractTxn(t, s, 2, 1)
+	b, err := mineTestingBlock(s.CurrentBlock().ID(), Timestamp(time.Now().Unix()), nullMinerPayouts(s.Height()+1), []Transaction{txn}, s.CurrentTarget())
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = s.AcceptBlock(b)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Submit 2 blocks, which means the contract will be a missed proof.
+	for i := 0; i < 2; i++ {
+		b, err = mineValidBlock(s)
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = s.AcceptBlock(b)
+		if err != nil {
+			t.Error(err)
+		}
+	}
+
+	// Check that the contract was removed, and that the missed proof output
+	// was added.
+	cid := txn.FileContractID(0)
+	_, exists := s.openContracts[cid]
+	if exists {
+		t.Error("file contract is still in state despite having terminated")
+	}
+	output, exists := s.unspentOutputs[cid.StorageProofOutputID(false)]
+	if !exists {
+		t.Error("missed storage proof output is not in state even though the proof was missed")
+	}
+
+	// Check that the money went to the right place.
+	if output.SpendHash != txn.FileContracts[0].MissedProofAddress {
+		t.Error("missed proof output sent to wrong address!")
+	}
+}
+
+// testStorageProofSubmit adds a block with a valid storage proof to the state
+// and checks that it is accepted, ending the contract supplying an output to
+// the person.
+func testStorageProofSubmit(t *testing.T, s *State) {
+	txn, cid := storageProofTxn(t, s)
+
+	// Get the contract for a later check. (Contract disappears after the block
+	// is accepted).
+	contract, exists := s.openContracts[cid]
+	if !exists {
+		t.Fatal("file contract doesn't exist in state")
+	}
+
+	b, err := mineTestingBlock(s.CurrentBlock().ID(), Timestamp(time.Now().Unix()), nullMinerPayouts(s.Height()+1), []Transaction{txn}, s.CurrentTarget())
```

Comment:
this appears in so many places (mostly tests) that I wonder if it's worth defining a convenience function:
```go
func currentTimestamp() Timestamp {
    return Timestamp(time.Now().Unix())
}
```